lms/services/vitalsource/service.py

```diff
@@ -148,3 +160,54 @@ def compile_user_lti_pattern(pattern: str) -> Optional[re.Pattern]:
             )
 
         return compiled_pattern
+
+    @property
+    def inclusive_access_enabled(self) -> bool:
+        """Check if the current school uses the inclusive access model to access hypothesis"""
+
+        return self._enable_inclusive_access
+
+    def inclusive_access_student_allowed(self, lti_params, course: Course):
+        user_reference = self.get_user_reference(lti_params)
+        vs_course_id = vitalsource_svc.get_vs_course_id(course)
+        if not vs_course_id:
+            # If VS is not aware of this course it means there's no opt outs
+            return
+        if vitalsource_svc.get_opt_out_for_user(vs_course_id, user_reference):
+            raise VitalSourceOptedOut()
+
+    def get_vs_course_id(self, course: Course):
+        if course_id := course.extra.get("vs", {}).get("course_id"):
+            return course_id
+
+        for vs_course in self._sso_client.get_courses():
+            if vs_course["lms_context_id"] == course.lms_id:
+                course.extra.setdefault("vs", {})
```

Comment:
I had this caching bit here, just on course but this would make sense probably also on "grouping_membership" for the student opt out deccison.

We'll have to decide
- Is this too slow to do in every launch?
- If it's indeed to slow we could:


    1. Follow the caching approach to speed thing up.
    2. Defer all of this to a celery task. A bit of delay of getting the data has the effect of allowing the student in instead of blocking.